src/components/cuda/cupti_common.c

```diff
@@ -0,0 +1,465 @@
+    cudaDriverGetVersionPtr    = DLSYM_AND_CHECK(dl_rt, "cudaDriverGetVersion");
+    cudaRuntimeGetVersionPtr   = DLSYM_AND_CHECK(dl_rt, "cudaRuntimeGetVersion");
+
+    Dl_info info;
+    dladdr(cudaGetDevicePtr, &info);
+    LOGDBG("CUDA runtime library loaded from %s\n", info.dli_fname);
+    return PAPI_OK;
+fn_fail:
+    return PAPI_EMISC;
+}
+
+static int unload_cudart_sym(void)
+{
+    if (dl_rt) {
+        dlclose(dl_rt);
+        dl_rt = NULL;
+    }
+    cudaGetDevicePtr           = NULL;
+    cudaGetDeviceCountPtr      = NULL;
+    cudaGetDevicePropertiesPtr = NULL;
+    cudaDeviceGetAttributePtr  = NULL;
+    cudaSetDevicePtr           = NULL;
+    cudaFreePtr                = NULL;
+    cudaDriverGetVersionPtr    = NULL;
+    cudaRuntimeGetVersionPtr   = NULL;
+    return PAPI_OK;
+}
+
+static int load_cupti_common_sym(void)
+{
+    char dlname[] = "libcupti.so";
+    char lookup_path[PATH_MAX];
+
+    char *papi_cuda_cupti = getenv("PAPI_CUDA_CUPTI");
+    if (papi_cuda_cupti) {
+        sprintf(lookup_path, "%s/%s", papi_cuda_cupti, dlname);
+        dl_cupti = dlopen(lookup_path, RTLD_NOW | RTLD_GLOBAL);
+    }
+
+    const char *standard_paths[] = {
+        "%s/extras/CUPTI/lib64/%s",
+        "%s/lib64/%s",
+        NULL,
+    };
+
+    char *papi_cuda_root = getenv("PAPI_CUDA_ROOT");
+    if (papi_cuda_root && !dl_cupti) {
+        dl_cupti = load_dynamic_syms(papi_cuda_root, dlname, standard_paths);
+    }
+
+    if (!dl_cupti) {
+        dl_cupti = dlopen(dlname, RTLD_NOW | RTLD_GLOBAL);
+        if (!dl_cupti) {
+            ERRDBG("Loading libcupti.so failed. Try setting PAPI_CUDA_ROOT\n");
+            goto fn_fail;
+        }
+    }
+
+    cuptiGetVersionPtr = DLSYM_AND_CHECK(dl_cupti, "cuptiGetVersion");
+
+    Dl_info info;
+    dladdr(cuptiGetVersionPtr, &info);
+    LOGDBG("CUPTI library loaded from %s\n", info.dli_fname);
+    return PAPI_OK;
+fn_fail:
+    return PAPI_EMISC;
+}
+
+static int unload_cupti_common_sym(void)
+{
+    if (dl_cupti) {
+        dlclose(dl_cupti);
+        dl_cupti = NULL;
+    }
+    cuptiGetVersionPtr = NULL;
+    return PAPI_OK;
+}
+
+static int util_load_cuda_sym(void)
+{
+    int papi_errno;
+    papi_errno = load_cuda_sym();
+    papi_errno += load_cudart_sym();
+    papi_errno += load_cupti_common_sym();
+    if (papi_errno != PAPI_OK) {
+        return PAPI_EMISC;
+    }
+    else
+        return PAPI_OK;
+}
+
+int util_unload_cuda_sym(void)
+{
+    unload_cuda_sym();
+    unload_cudart_sym();
+    unload_cupti_common_sym();
+    return PAPI_OK;
+}
+
+static int util_dylib_cu_runtime_version(void)
+{
+    int runtimeVersion;
+    CUDART_CALL(cudaRuntimeGetVersionPtr(&runtimeVersion), return PAPI_EMISC );
+    return runtimeVersion;
+}
+
+static int util_dylib_cupti_version(void)
+{
+    unsigned int cuptiVersion;
+    CUPTI_CALL(cuptiGetVersionPtr(&cuptiVersion), return PAPI_EMISC );
+    return cuptiVersion;
+}
+
+int get_device_count(void)
+{
+    static int numDevs = -1;
```

Comment:
why `-1`?

Reply:
`-1` means this function was not called before, any other value means it was called before and use that value instead. Have a look at the implementation where a static variable is used to track this.